[WIP] Add documentation for MTChannelBasedBroker

creydr opened the pull request:
Add an overview over the `MTChannelBasedBroker`.

aliok commented:
Diagram looks amazing!

We need something like that for EKB too!

Just one question about the diagram...

> (9)
> watches for new Subscriptions
> for the channel type and configures itself,
> so the dispatcher sends new events
> to the subscriptions subscriber
> (which points to the mt-broker-filter)

I don't remember exactly: does the dataplane watch for the Subscriptions? It talks to Kube API? This sounds very strange design now, when I think about the EKB dataplane which doesn't interact with Kube API.

aliok commented on docs/mt-channel-based-broker/README.md:
```diff
@@ -0,0 +1,56 @@
+# Multi-tenant channel-based broker
+
+This document gives a rough overview over the Multi-tenant channel-based broker (MTChannelBasedBroker) architecture.
```
```suggestion
This document gives a rough overview over the Multi-tenant channel-based broker (MTChannelBasedBroker) architecture.

It is possible to use MTChannelBasedBroker with [any channel implementation](https://knative.dev/docs/eventing/channels/channels-crds/). For demonstration purposes, MTChannelBasedBroker is used with an [InMemoryChannel](https://github.com/knative/eventing/blob/release-1.9/config/channels/in-memory-channel/README.md) in the diagram and throughout the document.
```

aliok commented on docs/mt-channel-based-broker/README.md:
```diff
@@ -0,0 +1,56 @@
+
+### Channel specific controllers (e.g. `imc-controller`)
+
+The channel specific controllers (`imc-controller` in our diagram) watch for channel resources created by the `mt-broker-controller` in step 3 (e.g. `InMemoryChannel`) and updates the address in the status pointing to the real endpoint of the channel (a service backed by the `imc-dispatcher` and not via the `mt-broker-ingress`).
```
```suggestion
As we are using `InMemoryChannel` as the backing channel for the broker throughout this document, the channel specific controllers will be related to that implementation.

These controllers (`imc-controller` in our diagram) watch for channel resources created by the `mt-broker-controller` in step 3 (e.g. `InMemoryChannel`) and updates the address in the status pointing to the real endpoint of the channel (a service backed by the `imc-dispatcher` and not via the `mt-broker-ingress`).
```

creydr commented:
> We need something like that for EKB too!

On my list :)

> I don't remember exactly: does the dataplane watch for the Subscriptions? It talks to Kube API? This sounds very strange design now, when I think about the EKB dataplane which doesn't interact with Kube API.

Ah you're are rigth. That was a mistake. Thanks for catching this. It seems the subscription controller updates the channel with the subscribers IIUC (step 9) and the channel-specific-data-plane component reconciles the channel which then contains the subscribers (step 10).
Just to be sure: can you confirm this @aliok @pierDipi @matzew

pierDipi commented:
That is correct @creydr

creydr commented:
Follow up would be to reference this in the docs (https://knative.dev/docs/eventing/brokers/broker-types/)

aliok commented:
/lgtm
/approve

pierDipi reviewed (commented):
LGTM, there are a few linting errors that should be fixed before merging

creydr commented:
> @creydr Fantastic work with the diagram (I especially like your way of showing control and data plane). I would add .drawio extension to the file to make it easier to open.

Thank you @aslom. @pierDipi gave me a lot of context for this :)
I updated the file name in f4fe244422df64c098a1e462975ecdc2d33fb754

> there are a few linting errors that should be fixed before merging

@pierDipi: I addressed them in 017d40fdce9bcb23c303dd86e86c66bcd7f7144f

pierDipi reviewed (approved):
/lgtm
/approve

creydr commented:
/hold cancel
